elastio-nat-provision-lambda/lambda.py

```diff
@@ -0,0 +1,348 @@
+def request(client, operation, query, **kwargs):
+    """
+    Sends a paginated request to AWS, filters the response with a JMESPath
+    query and returns a list.
+    """
+    paginator = client.get_paginator(operation)
+    page_iter = paginator.paginate(**kwargs)
+    filtered = page_iter.search(query)
+    return list(filtered)
+
+
+def ensure_nat(instance_id):
+    instance = request(
+        ec2,
+        'describe_instances',
+        'Reservations[].Instances[]',
+        InstanceIds=[instance_id],
+    )[0]
+
+    print("instance:", instance)
+
+    instance_vpc_id = instance['VpcId']
+    instance_subnet_id = instance['SubnetId']
+    instance_tags = instance['Tags']
+
+    if not any(
+            tag['Key'] == 'elastio:resource' and tag['Value'] == 'true'
+            for tag in instance_tags
+    ):
+        print(f"No matching elastio:resource tag found on instance {instance_id}; no action taken.")
+        return
+
+    subnets = {sn['SubnetId']: sn for sn in request(
+        ec2,
+        'describe_subnets',
+        'Subnets',
+        Filters=[{'Name': 'vpc-id', 'Values': [instance_vpc_id]}],
+    )}
+
+    route_tables = {rt['RouteTableId']: rt for rt in request(
+        ec2,
+        'describe_route_tables',
+        'RouteTables',
+        Filters=[{'Name': 'vpc-id', 'Values': [instance_vpc_id]}],
+    )}
+
+    main_route_table_id = None
+    subnet_to_route_table = {}
+
+    for rt_id, rt in route_tables.items():
+        for assoc in rt['Associations']:
+            if assoc['Main']:
+                main_route_table_id = rt_id
+            subnet_id = assoc.get('SubnetId')
+            if subnet_id:
+                subnet_to_route_table[subnet_id] = rt_id
+
+    for subnet_id in subnets:
+        if subnet_id not in subnet_to_route_table:
+            subnet_to_route_table[subnet_id] = main_route_table_id
+
+    public_subnets_ids = set(get_public_subnets(subnet_to_route_table, route_tables))
+    instance_route_table_id = subnet_to_route_table[instance_subnet_id]
+    instance_route_table = route_tables[instance_route_table_id]
+
+    if not public_subnets_ids:
+        print(f"No public subnets found in {instance_vpc_id}; exiting")
+        return
+
+    if instance_subnet_id in public_subnets_ids:
+        print("Instance is running in a public subnet; exiting")
+        return
+
+    all_traffic_route = get_all_traffic_route(instance_route_table)
+
+    if all_traffic_route is not None:
+        print(f"Route table already has a route for 0.0.0.0/0; exiting. {all_traffic_route}")
+        return
+
+    print("subnets:", subnets)
+    print("public_subnets_ids:", public_subnets_ids)
+    print("instance_subnet_id:", instance_subnet_id)
+
+    nat_subnet_id = choose_subnet_for_nat(subnets, public_subnets_ids, instance_subnet_id)
+
+    if nat_subnet_id is None:
+        print("Unable to find a subnet in the same availability zone; exiting")
+        return
+
+    stack_name = f"{NAT_CFN_PREFIX}{instance_vpc_id}"
```

Comment:
If there are two instances, each spun up in a different subnet, this will break won't it? You should not assume that only one subnet is enabled for the vault deployment.  The customer we have initially created this mechanism for typically deploys into two subnets for example.